mariadb*: apply upstream patch and fix for Linux

danielnachun opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

danielnachun commented:
I've found the upstream commit which should fix the CMake error that has been blocking all versions of `mariadb` from being built.  The problem is actually in `mariadb-connector-c`, which hasn't tagged a new release with this commit yet.

carlocab commented:
Still suffers from the `#include <version>` bug.
```
In file included from /tmp/mariadb-20210621-28458-km2ph1/mariadb-10.5.9/storage/perfschema/cursor_by_account.cc:28:
In file included from /tmp/mariadb-20210621-28458-km2ph1/mariadb-10.5.9/include/my_global.h:287:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/math.h:309:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:417:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:37:
/tmp/mariadb-20210621-28458-km2ph1/mariadb-10.5.9/version:1:1: error: expected unqualified-id
MYSQL_VERSION_MAJOR=10
^
/tmp/mariadb-20210621-28458-km2ph1/mariadb-10.5.9/include/my_config.h:527:29: note: expanded from macro 'MYSQL_VERSION_MAJOR'
#define MYSQL_VERSION_MAJOR 10
                            ^
```

danielnachun commented:
It is claimed by upstream that the latest point releases of all versions of MariaDB they support should have this fixed: https://jira.mariadb.org/browse/MDEV-25543.  That may also fix the CMake error, removing the need for applying the upstream patch.  I will test locally and see if moving to the next version is sufficient.

cho-m commented:
@danielnachun See #76887 for attempt at updating to 10.5.10

Maybe one of newer releases fixes all the issues I noted: https://github.com/Homebrew/homebrew-core/pull/76887#issuecomment-840851149

danielnachun commented:
Unfortunately the new releases do not fix these issues.  There is a very recent commit which I was able to use the fix the issue with S3, but it is not in the most recent release: https://github.com/mariadb-corporation/libmarias3/commit/c71898f8259829bf4b2e01072000491281e5206d.  

There is an old issue for the `mroonga` problem which was never resolved: https://jira.mariadb.org/browse/MDEV-7209?jql=text%20~%20%22expected%20unqualified%22.  It seems this issue is a problem for BSD systems in general, not just macOS, as that issue was reported in OpenBSD.  I think we could make a comment in that issue letting them know the real cause of the problem, as they mistakenly thought it was just because `mroonga` was too old.

This is a bit of bind for us on the Linux side right now, as with the newest stable releases, all versions of MariaDB build with no patching as long as we use GCC-11.  But clearly on the macOS side there are some issues that upstream needs to fix before it will work without patching.  

Perhaps if we update that upstream issue and link it, we can combine the fix @cho-m made for `mroonga` with the upstream commit to fix S3 and that would be acceptable?

cho-m commented:
> It seems this issue is a problem for BSD systems in general, not just macOS, as that issue was reported in OpenBSD

Should be related to compiling with a (more-recent?) LLVM Clang compiler (not sure if any other C/C++ compiler impacted).

The problem with macOS is the case-insensitive filesystem makes the problem worse by breaking on `VERSION` files too.

---

> I think we could make a comment in that issue letting them know the real cause of the problem, as they mistakenly thought it was just because `mroonga` was too old.

Probably a good idea so that we have an upstream issue to reference. Could also link the other MariaDB issues to show how it was fixed in other places.

---

> Perhaps if we update that upstream issue and link it, we can combine the fix @cho-m made for `mroonga` with the upstream commit to fix S3 and that would be acceptable?

May need to tweak some of the fixes (perhaps renaming would be better than removing files), but using local patches are going to be the fastest way for us to fix Formula for both macOS 11.3+ and Linux.

Hard to tell how long an upstream fix would take.

As long as we comment the temporary fixes and reference corresponding upstream issues, we should be fine locally patching/fixing.

For existing S3 patch, it may still require an additional `File.rename` as I see some other Formula needed to do this even with a diff that does rename the file.

danielnachun commented:
I just pushed what I think are the necessary fixes for all versions to build successfully, based on a lot of local testing.  A few key points:

1) I bumped everything to the latest versions, all of which were released on 6/23.
2) I adapted the fixes that @cho-m made, with the one difference being that I renamed the version/VERSION files instead of deleting them.  
3) The problem with S3 only affects 10.5, so that fix is not needed for the older versions. 

Once the CI for this is green, I'm going to post a comment in the Jira issue I mentioned above explaining how we've fixed this.  If upstream doesn't respond, I will see if I can open a pull request, as these fixes are very simple.

danielnachun commented:
Now that this is all green, I have made posts to the following upstream issues:

libmarias3: https://jira.mariadb.org/browse/MDEV-25716
mroonga: https://jira.mariadb.org/browse/MDEV-7209

Let me know if this should linked in the comments in the formulae.

SMillerDev reviewed (commented):
LGTM

danielnachun commented:
Upstream has responded to my comments.  I forgot that the `libmarias3` issue is fixed in an upstream commit, it just didn't make it into 10.5.11.  It should be in the next stable release of 10.5.  They have assigned someone to fix mroonga as well.  So hopefully both of these will be fully resolved in the next set of stable releases.

carlocab commented on Formula/mariadb.rb:
```diff
@@ -1,8 +1,8 @@
   homepage "https://mariadb.org/"
-  url "https://downloads.mariadb.com/MariaDB/mariadb-10.5.9/source/mariadb-10.5.9.tar.gz"
-  sha256 "40ab19aeb8de141fdc188cf2251213c9e7351bee4d0cd29db704fae68d1068cf"
+  url "https://downloads.mariadb.com/MariaDB/mariadb-10.5.11/source/mariadb-10.5.11.tar.gz"
```
Can we add a comment here about removing the patch and the `mv` workarounds in the next version (or whenever it is they can be removed)?

cho-m commented:
In terms of build, I think it is fine.

Adding some comments relating to temporary fixes would be good for future update.
Can either do it here or in a separate syntax-only PR.

Looks like Linux bottle in `homebrew-core` would need `groonga` to be bottled first.

danielnachun commented:
I've added comments to all versions explaining the fixes.

I'd favor handling the Linux bottling in a separate PR because this one was very complicated as is.

danielnachun commented:
The only reason this failed is because of a flaky upstream URL.  The Mojave bottle for 10.3 did build.  

I should have asked this much earlier, but do we need a revision bump for this?

carlocab reviewed (approved):
These are all version bumps, so a revision bump isn't necessary.
